keps/sig-node/20200327-slim-down-kubelet-and-its-functions.md

```diff
@@ -0,0 +1,295 @@
+---
+title: Slim Down Kubelet and its functions
+authors:
+  - "@gongguan"
+  - TBD
+owning-sig: sig-node
+reviewers:
+  - "@dashpole"
+  - "@derekwaynecarr"
+approvers:
+  - "@dchen1107"
+  - "@derekwaynecarr"
+creation-date: 2020-03-27
+last-updated: 2020-04-11
+status: provisional
+---
+
+# Slim Down Kubelet and its functions
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Pros](#pros)
+  - [Cons](#cons)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Better modularize Kubelet](#better-modularize-kubelet)
+  - [Slim down Kubelet functions](#slim-down-kubelet-functions)
+  - [Test Migration](#test-migration)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+Current kubelet root directory is too large to extend and components dependency blurry.
+In current Kubelet, all the members are put into it frankly. We should split it up by better modularized.
+Kubelet was injected to some components like `kubeGenericRuntimeManager`, `resourceAnalyzer` and `VolumePluginMgr` which caused many methods implemented in Kubelet struct. We'd better converge them to its sub components.
+
+## Motivation
+
+Better modularize Kubelet and converge many Kubelet functions to its sub components which slim down both Kubelet and its functions.
+
+### Pros
+- Make kubelet(very huge now) better Modularized.
+- Slim down Kubelet functions by returning many functions to kubeGenericRuntimeManager or resourceAnalyzer.
+- Resolve some circular dependencies like kubeGenericRuntimeManager and volumeManager, StatsProvider and ResourceAnalyzer.
+
+Kubelet components dependency states and migration plan will be listed in Proposal.
+
+### Cons
+I think this migration with low risk, because functions won't be changed during migration.
+
+### Goals
+- Not inject Kubelet to its components' initialization which makes components dependencies clear.
+- Slim down Kubelet structure and its functions.
+
+### Non-Goals
+- Any effect on Kubelet current function.
+
+## Proposal
+
+### Better modularize Kubelet
+
+Create `NodeInfo`(pkg/kubelet/nodeinfo) and `DirectoryManager`(pkg/kubelet/directorymanager) only contains basic members which often used by kubelet components.
+
+```go
+// NodeInfo's members are migrated from Kubelet.
+type NodeInfo struct {
```

Comment:
can we make this struct, and fields in it private, since we have a public interface?

Reply:
Sure, done. Can we not add functions such as GetHostName() to interface in the kep, these just make the kep tedious.

Reply:
I suppose thats fine.  At least make a note of it somewhere, e.g. `GetFoo() // for constants in nodeInfo above`

Reply:
Done.